Drag and Drop

crisbeto opened the pull request:
Live demo: https://material-cb7ec.firebaseapp.com/drag-and-drop

crisbeto commented:
CC @jelbourn

jelbourn commented on src/demo-app/drag-and-drop/drag-and-drop-demo.ts:
```diff
@@ -0,0 +1,499 @@
+import {DomSanitizer} from '@angular/platform-browser';
+
+/** Event emitted when the user starts dragging an item in a drop container. */
+interface CdkDropStart<T> {
```
I'll put all my thoughts on event naming in one place, based on the native drag/drop API:

I'm thinking the set of event classes should be:
`CdkDragStart`: fired by a draggable when it starts being dragged.
`CdkDragEnd`: fired by a draggable when it stops being dragged.
`CdkDragEnter`: fired when a draggable enters a dropzone. Can be fired by either the draggable or the drop zone. Event includes both instances.
`CdkDragExit`: fired when a draggable leaves a dropzone. Can be fired by either the draggable or the drop zone. Event includes both instances.
`CdkDragDrop`: fired when a dragging ends inside a drop zone. Can be fired by either the draggable or the drop zone. Event includes both instances.

* I think we want to avoid having an equivalent to `dragover` because we wouldn't want change detection to run that much.
* I'm not sure how the "transfer" or "remove" events fit into this; is there a reason they couldn't fit into one of the other events above?

crisbeto replied:
I added the `transfer` and `remove` events so the consumer has a way of knowing when to remove the item from one array and add it to the other. We need these to fire once the user is done dragging, otherwise the consumer might swap out the data in an `ngFor` which will destroy the `CdkDrag` instance that is currently active. Are `CdkDragEnter` and `CdkDragExit` supposed to fire while the user is dragging or when they've dropped the item?

Also, technically the `remove` event isn't necessary since the `transfer` event has the data of both the source and destination drop containers. I put it in for consistency.

jelbourn replied:
* It sounds like `transfer` is just what I'm calling `drop`, except that it also happens when its within the same zone. Is there any reason you could do the remove on `drop`? As long as the user updates both data sources for the source and target locations in the same tick it should be okay.
* `enter` and `exit` would fire while you're still dragging

crisbeto replied:
What I meant is that the remove should only happen on `drop`, otherwise we risk it breaking while the user is still dragging, but I think that's what we agreed on previously.

jelbourn replied:
Yeah, agreed. Though we should try to gracefully fail when the user destroys the draggable in the middle of a drag.

jelbourn reviewed (commented):
Overall looks good, I just have some naming suggestions and questions about the events. I ran the overall concept by Igor and he's also on board.

I think next steps are:
1. Create a checklist of work left to do in the main drag/drop issue (you can edit my top comment). Off the top of my head, some of the things this would include are:
* Reordering within a list
* Dragging elements between different lists
* Free moving an element
* Multi-touch (dragging two things at the same time)
* Nested drop zones
* Documentation
* Accessibility instructions

And anything else you can think of

2. Start sending PRs into cdk-experimental

crisbeto committed:
feat: add experimental drag-drop package

Adds an initial version of the `drag-drop` module to `cdk-experimental` with support for the following:

* Dragging an element on its own.
* Dragging an element using a handle.
* Sorting an element within a drop zone.
* Moving an element between drop zones.
* Touch and mouse support.
* Custom template for the preview element.
* Custom template for the placeholder element.
* Animation support for the preview.

Demo: https://material-cb7ec.firebaseapp.com/drag-drop

Relates to #8963.
